Added support for C#, VB & F# Languages with .NET Standard 1.0

DJGosnell opened the pull request:
I have completed the initial porting of the JavaScript version of the library to .NET Standard in C#.

The only change I have made in the operation of the system is to put all the functions in an instanceable class which requires the unit system as a constructor parameter.  This is to allow simultaneous SI & IP calculations on different instances which has the added benefit of allowing all the methods to be thread-safe.  
I have updated the [examples](https://github.com/DJGosnell/psychrolib/blob/master/docs/examples.md) to include C# language usage. 

For those of you not familiar with the .NET ecosystem, .NET Standard 1.0 is a library format used by the .NET Framework >= 4.5, .NET Core >= 1 and others.  .net Core supports Windows, Linux & Mac, so this library is also cross platform.
[Compatibility Reference](https://docs.microsoft.com/en-us/dotnet/standard/net-standard)

I have ported the complete [unit test suite](https://github.com/DJGosnell/psychrolib/tree/master/src/c_sharp/PsychroLib.Tests) and have the results posted below.
```
PsychroLib.Tests (100 tests) [0:00.844] Success
 PsychroLib.Tests (100 tests) [0:00.844] Success
  PsychroLibUnitTests (100 tests) [0:00.844] Success
   AllPsychrometrics [0:00.021] Success
   DryAir (2 tests) [0:00.001] Success
    DryAir(IP,77,18.498d,14.696d,13.5251d,0.07393660675336966d,42.6168d,0.02d,86,0.001d) [0:00.001] Success
    DryAir(SI,25,25148,101325,0.8443d,1.184413123297406d,81316,0.02d,30,0.001d) [0:00.000] Success
   GetHumRatioFromSpecificHum (2 tests) [0:00.000] Success
    GetHumRatioFromSpecificHum(IP,0.00596421471d,0.006d,0.01d) [0:00.000] Success
    GetHumRatioFromSpecificHum(SI,0.00596421471d,0.006d,0.01d) [0:00.000] Success
   GetSatAirEnthalpy (14 tests) [0:00.001] Success
    GetSatAirEnthalpy(IP,-4,14.696d,-0.286d,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(IP,-58,14.696d,-13.906d,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(IP,122,14.696d,126.066d,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(IP,185,14.696d,999.749d,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(IP,23,14.696d,8.186d,0.03d) [0:00.000] Success
    GetSatAirEnthalpy(IP,41,14.696d,15.699d,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(IP,77,14.696d,40.576d,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(SI,-20,101325,-18542,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(SI,-5,101325,1164,0.03d) [0:00.000] Success
    GetSatAirEnthalpy(SI,-50,101325,-50222,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(SI,25,101325,76504,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(SI,5,101325,18639,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(SI,50,101325,275353,0.01d) [0:00.000] Success
    GetSatAirEnthalpy(SI,85,101325,2307539,0.01d) [0:00.000] Success
   GetSatHumRatio (14 tests) [0:00.002] Success
    GetSatHumRatio(IP,-4,14.696d,0.0006373d,0.01d) [0:00.000] Success
    GetSatHumRatio(IP,-58,14.696d,2.43E-05d,0.01d) [0:00.000] Success
    GetSatHumRatio(IP,122,14.696d,0.086863d,0.01d) [0:00.000] Success
    GetSatHumRatio(IP,185,14.696d,0.838105d,0.02d) [0:00.000] Success
    GetSatHumRatio(IP,23,14.696d,0.0024863d,0.005d) [0:00.000] Success
    GetSatHumRatio(IP,41,14.696d,0.005425d,0.005d) [0:00.000] Success
    GetSatHumRatio(IP,77,14.696d,0.020173d,0.005d) [0:00.000] Success
    GetSatHumRatio(SI,-20,101325,0.0006373d,0.01d) [0:00.000] Success
    GetSatHumRatio(SI,-5,101325,0.0024863d,0.005d) [0:00.000] Success
    GetSatHumRatio(SI,-50,101325,2.43E-05d,0.01d) [0:00.000] Success
    GetSatHumRatio(SI,25,101325,0.020173d,0.005d) [0:00.000] Success
    GetSatHumRatio(SI,5,101325,0.005425d,0.005d) [0:00.000] Success
    GetSatHumRatio(SI,50,101325,0.086863d,0.01d) [0:00.000] Success
    GetSatHumRatio(SI,85,101325,0.838105d,0.02d) [0:00.000] Success
   GetSatVapPres_Absolute (2 tests) [0:00.000] Success
    GetSatVapPres_Absolute(IP,-76,0.000157d,1E-05.0d) [0:00.000] Success
    GetSatVapPres_Absolute(SI,-60,1.08d,0.01d) [0:00.000] Success
   GetSatVapPres_Relative (14 tests) [0:00.002] Success
    GetSatVapPres_Relative(IP,-4,0.014974d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(IP,122,1.7914d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(IP,212,14.7094d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(IP,23,0.058268d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(IP,300,67.0206d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(IP,41,0.12656d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(IP,77,0.45973d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,-20,103.24d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,-5,401.74d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,100,101418.0d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,150,476101.4d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,25,3169.7d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,5,872.6d,0.0003d) [0:00.000] Success
    GetSatVapPres_Relative(SI,50,12351.3d,0.0003d) [0:00.000] Success
   GetSpecificHumFromHumRatio (2 tests) [0:00.000] Success
    GetSpecificHumFromHumRatio(IP,0.006d,0.00596421471d,0.01d) [0:00.000] Success
    GetSpecificHumFromHumRatio(SI,0.006d,0.00596421471d,0.01d) [0:00.000] Success
   GetStandardAtmPressure (12 tests) [0:00.001] Success
    GetStandardAtmPressure(IP,-1000,15.236d,1) [0:00.000] Success
    GetStandardAtmPressure(IP,0,14.696d,1) [0:00.000] Success
    GetStandardAtmPressure(IP,1000,14.175d,1) [0:00.000] Success
    GetStandardAtmPressure(IP,10000,10.108d,1) [0:00.000] Success
    GetStandardAtmPressure(IP,3000,13.173d,1) [0:00.000] Success
    GetStandardAtmPressure(IP,30000,4.371d,1) [0:00.000] Success
    GetStandardAtmPressure(SI,-500,107478,1) [0:00.000] Success
    GetStandardAtmPressure(SI,0,101325,1) [0:00.000] Success
    GetStandardAtmPressure(SI,1000,89875,1) [0:00.000] Success
    GetStandardAtmPressure(SI,10000,26436,1) [0:00.000] Success
    GetStandardAtmPressure(SI,4000,61640,1) [0:00.000] Success
    GetStandardAtmPressure(SI,500,95461,1) [0:00.000] Success
   GetStandardAtmTemperature (12 tests) [0:00.001] Success
    GetStandardAtmTemperature(IP,-1000,62.6d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(IP,0,59.0d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(IP,1000,55.4d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(IP,10000,23.4d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(IP,3000,48.3d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(IP,30000,-47.8d,0.2d) [0:00.000] Success
    GetStandardAtmTemperature(SI,-500,18.2d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(SI,0,15.0d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(SI,1000,8.5d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(SI,10000,-50.0d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(SI,4000,-11.0d,0.1d) [0:00.000] Success
    GetStandardAtmTemperature(SI,500,11.8d,0.1d) [0:00.000] Success
   GetTDewPointFromVapPres_convergence (2 tests) [0:00.804] Success
    GetTDewPointFromVapPres_convergence(IP,-148,392,1,8.6d,17.4d,1) [0:00.593] Success
53559 total iterations.
    GetTDewPointFromVapPres_convergence(SI,-100.0d,200.0d,1,60000,120000,10000) [0:00.210] Success
23177 total iterations.
   GetTKelvinFromTCelsius [0:00.000] Success
   GetTRankineFromTFahrenheit [0:00.000] Success
   HumRatio_TWetBulb (4 tests) [0:00.000] Success
    HumRatio_TWetBulb(IP,30.2d,23,14.175d,0.00114657481090184d,0.0003d,0.001d) [0:00.000] Success
    HumRatio_TWetBulb(IP,86,77,14.175d,0.0187193288418892d,0.0003d,0.001d) [0:00.000] Success
    HumRatio_TWetBulb(SI,-1,-5,95461,0.00120399819933844d,0.0003d,0.001d) [0:00.000] Success
    HumRatio_TWetBulb(SI,30,25,95461,0.0192281274241096d,0.0003d,0.001d) [0:00.000] Success
   HumRatio_TWetBulbClamp (2 tests) [0:00.000] Success
    HumRatio_TWetBulbClamp(IP,23) [0:00.000] Success
    HumRatio_TWetBulbClamp(SI,-5) [0:00.000] Success
   HumRatio_VapPres (2 tests) [0:00.001] Success
    HumRatio_VapPres(IP,0.45973d,14.175d,0.0208473311024865d,1E-06.0d,1E-05.0d) [0:00.001] Success
    HumRatio_VapPres(SI,3169.7d,95461,0.0213603998047487d,1E-06.0d,1E-05.0d) [0:00.000] Success
   MoistAir (2 tests) [0:00.000] Success
    MoistAir(IP,86,0.02d,42.6168d,14.175d,14.7205749002918d,0.0692907720594378d,0.0003d) [0:00.000] Success
    MoistAir(SI,30,0.02d,81316,95461,0.940855374352943d,1.08411986348219d,0.0003d) [0:00.000] Success
   SeaLevel_Station_Pressure (2 tests) [0:00.000] Success
    SeaLevel_Station_Pressure(IP,14.681662559d,344.488d,62.942d,14.8640475d,0.0001d) [0:00.000] Success
    SeaLevel_Station_Pressure(SI,101226.5d,105,17.19d,102484.0d,1) [0:00.000] Success
   TWetBulb_RelHum (1 test) [0:00.000] Success
    TWetBulb_RelHum(SI,7,0.61d,100000,3.92667433781955d,0.001d) [0:00.000] Success
   VapPres_RelHum (2 tests) [0:00.000] Success
    VapPres_RelHum(IP,77,0.8d,0.36778400000000006d,0.8d,0.0003d) [0:00.000] Success
    VapPres_RelHum(SI,25,0.8d,2535.76d,0.8d,0.0003d) [0:00.000] Success
   VapPres_TDewPoint (6 tests) [0:00.000] Success
    VapPres_TDewPoint(IP,-4.0d,59.0d,-4.0d,0.001d) [0:00.000] Success
    VapPres_TDewPoint(IP,122.0d,140.0d,122.0d,0.001d) [0:00.000] Success
    VapPres_TDewPoint(IP,41.0d,59.0d,41.0d,0.001d) [0:00.000] Success
    VapPres_TDewPoint(SI,-20.0d,15.0d,-20.0d,0.001d) [0:00.000] Success
    VapPres_TDewPoint(SI,5.0d,15.0d,5.0d,0.001d) [0:00.000] Success
    VapPres_TDewPoint(SI,50.0d,60.0d,50.0d,0.001d) [0:00.000] Success
```

dmey commented:
@DJGosnell well thanks and many thanks for your contribution! 👍 

I have just added myself and @didierthevenard as reviewers. I have not looked at this carefully but I noticed that you have included solution files for VS. Are you familiar with CMake? I would prefer to use that to generate the solution files where possible -- otherwise let me know and I can have a look. There is also the CI to set up and I guess we may need/want to start adding Windows but again I am happy to have a look at that .

@didierthevenard let me know what you think as well and if you have any major comments before starting with the review.

@didierthevenard @DJGosnell just to let you know I am a bit overstretched these next few weeks so I may be a bit slow in getting back.

DJGosnell commented:
@dmey I have not looked at using CMake for .NET projects in the past as I have mainly been a Windows developer, with a medium amount of experience in Linux.  I will have to do some investigation of this.

.NET Core has broken a lot of stigmas associated with Microsoft and cross platform accessibility.  In the past, we had the Mono project to help with this, but since the .NET community has been focused on .NET Core and associated projects, other OSes are now fully supported by Microsoft itself.

With all that said, sln files are no longer only utilized by Visual Studio.  Once the .NET Core SDK has been installed on your current flavor of OS (Linux, Windows, Mac), you can simply run "dotnet build" or "dotnet test" utilizing the sln and associated .csproj files.  As such, I do not believe that there is any need to any anything specifically for Windows over other operating systems.

If CMake is a requirement, I will do further investigation, but it appears to me that the built-in methods in the .NET Core packages handle the old issues caused by using Visual Studio.

I am investigating on the best method to handle CI integration right now.

As far as timing, I completely understand.  I did this initial port for a work project which required it, but now that has been completed, I will invest what off-time I have to address these other items.

DJGosnell commented:
CI has been updated to include C# project.

dmey commented on README.md:
```diff
@@ -5,7 +5,7 @@
 [![Build Status](https://travis-ci.com/psychrometrics/psychrolib.svg?branch=master)](https://travis-ci.com/psychrometrics/psychrolib) | [![DOI](https://joss.theoj.org/papers/10.21105/joss.01137/status.svg)](https://doi.org/10.21105/joss.01137) | [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2537945.svg)](https://doi.org/10.5281/zenodo.2537945)|
 
-PsychroLib is a library of functions to enable the calculation of psychrometric properties of moist and dry air. Versions in Python, C, Fortran, JavaScript, and Microsoft Excel Visual Basic for Applications (VBA) are available. The library works in both metric (SI) and imperial (IP) systems of units. For a general overview and a list of currently available functions, please see the [overview page ...
+PsychroLib is a library of functions to enable the calculation of psychrometric properties of moist and dry air. Versions in Python, C, Fortran, JavaScript, Microsoft Excel Visual Basic for Applications (VBA), C#, Visual Basic, and F# with .NET Standard 1.0 (.NET Core >= 1.0 & .NET Framework >= 4.5) are available. The library works in both metric (SI) and imperial (IP) systems of units. For a gen ...
```
```suggestion
PsychroLib is a library of functions to enable the calculation of psychrometric properties of moist and dry air. Versions are available in Python, C, Fortran, JavaScript, Microsoft Excel Visual Basic for Applications (VBA), and C#, Visual Basic, and F# with .NET Standard 1.0 (.NET Core >= 1.0 & .NET Framework >= 4.5). The library works in both metric (SI) and imperial (IP) systems of units. For a general overview and a list of currently available functions, please see the [overview page](docs/overview.md).
```

dmey commented on README.md:
```diff
@@ -23,6 +23,8 @@ Examples on how to use PsychroLib in all the supported languages are described i
 
 - C, Fortran and JavaScript: clone the repository, and bundle according to your requirements.
 
+- C#: Use the [NuGet package](https://www.nuget.org/packages/PsychroLib/). -OR- Clone the repository, and bundle according to your requirements.
```
```suggestion
- C#: from the [NuGet package](https://www.nuget.org/packages/PsychroLib/) manager or clone the repository, and bundle according to your requirements.
```

dmey commented on docs/overview.md:
```diff
@@ -4,7 +4,7 @@ Psychrometrics are the study of physical and thermodynamic properties of moist a
 The estimation of these properties is critical in several engineering and scientific applications such as heating, ventilation, and air conditioning (HVAC) and meteorology. Although formulae to calculate the psychrometric properties of air are widely available in the literature ([@Stull2011]; [@Wexler1983]; [@Stoecker1982]; [@Dilley1968]; [@Humphreys1920]), their implementation in computer progra ...
 
-PsychroLib is a library of functions to enable calculating psychrometric properties of moist and dry air. The library is available for Python, C, Fortran, JavaScript, and Microsoft Excel Visual Basic for Applications (VBA). It works in both metric (SI) and imperial (IP) systems of units. The functions are based of formulae from the  2017 ASHRAE Handbook — Fundamentals, Chapter 1, SI and IP editio ...
+PsychroLib is a library of functions to enable calculating psychrometric properties of moist and dry air. The library is available for Python, C, Fortran, JavaScript, Microsoft Excel Visual Basic for Applications (VBA), C#, Visual Basic, and F# with .NET Standard 1.0 (.NET Core >= 1.0 & .NET Framework >= 4.5). It works in both metric (SI) and imperial (IP) systems of units. The functions are base ...
```
```suggestion
PsychroLib is a library of functions to enable calculating psychrometric properties of moist and dry air. The library is available for Python, C, Fortran, JavaScript, Microsoft Excel Visual Basic for Applications (VBA), and C#, Visual Basic, and F# with .NET Standard 1.0 (.NET Core >= 1.0 & .NET Framework >= 4.5). It works in both metric (SI) and imperial (IP) systems of units. The functions are based of formulae from the  2017 ASHRAE Handbook — Fundamentals, Chapter 1, SI and IP editions. Functions can be grouped into two categories:
```

dmey commented on src/c_sharp/PsychroLib/Psychrometrics.cs:
```diff
@@ -0,0 +1,1080 @@
+﻿/*
+ * PsychroLib (version 2.1.1) (https://github.com/psychrometrics/psychrolib)
+ * Copyright (c) 2018 D. Thevenard and D. Meyer for the current library implementation
```
I think you should add your initials and name given your work!

dmey commented:
@DJGosnell many thanks. I have just sent you some comments re your PR. Re CMake, I'm happy with the current setup and solution files -- as you suggested it works well and it's cross-platform!

DJGosnell commented:
Requested modifications by @dmey are complete.

1. Done
2. Done
3. Done
4. You are correct, this is non-standard for C#, but these modifications have been applied.
5. Updated version.  Waiting until PR is completed until updating NuGet package.
6. I do not mind maintaining the C# portion as long as there is an understanding that there may be delays in responses at times.

dmey commented:
@DJGosnell looks good, thanks!

@didierthevenard I am happy with this PR, when you get a chance let me know if you are also happy with it and I will merge it in and release a new version (2.2.0).

didierthevenard commented:
Hello @DJGosnell and @dmey ,

Apologies for the late reply, I am on sabbatical leave and just starting to touch the keyboard again.
Thank you @DJGosnell for your port to C#, I think it is great to add that language and it looks really good. I will let you and @dmey handle the metacode files, I just looked at the C# code itself. I did a line-by-line comparison to the C version and I think you have done a very thorough job, everything looks great, except for one bug which is _not_ specific to C# and a couple of minor issues.

**Minor issues:**

1. You named your class Phychrometrics, I assume you wanted to call it Psychrometrics?
2. I am not familiar with the C# syntax, but I am wondering what is the purpose of having 'public UnitSystem UnitSystem'. Is that just a function to initialize or retrieve the _unitSystem private member?
3. As you explained, you encapsulated the library into a class whose constructor initializes the _unitSystem private member. I think this is great (particularly for the reasons you mention). Again I am not familiar with the C# language, but is there any possibility that the class can be initialized without calling the constructor - that is, is there a default constructor that would leave _unitSystem uninitialized? I am asking because then this could lead to issues in some functions; please let me know if this is the case.
4. You seem to use camelCase notation, which is why TDryBulb becomes tDryBulb, which is fine by me. But then I notice that it is not used consistently, for example Wsstar in GetHumRatioFromTWetBulb, or SatVaporPres and SatHumRatio in GetSatHumRatio(), or BoundedHumRatio in various functions, don't follow the same convention. Is there any reason for that?
5. In the comments before GetSatVapPres and dLnPws_ I see the symbol '&amp;', is that on purpose or is it a text conversion issue? Is '&' not allowed in comments in C#?
6. It's a minor point, but in other languages there are 'section headings' before big chunks of code, like 'Global constants', 'Conversion between temperature units', etc. You have kept some but eliminated others, like the ones I mention above or 'Standard atmosphere'. Is that on purpose?

**Bug:**

7. In dLnPws_, in the IP case below the triple point of water, I believe that the coefficient in front of the fourth term on the right-hand side should be a 3, not a 2, because it originates in the derivation of T**3 (see the formula that this is derived from in GetSatVapPres()). A quick look leads me to believe that this issue is common to _all languages_ in PsychroLib. @dmey , can you please confirm my analysis? If so we need to fix it - let me know which languages you can take care of and which ones you want me to do.

dmey commented:
@didierthevenard well spotted! You are correct -- the coefficient from the first derivative should have been 3 and not 2. I have fixed this in https://github.com/psychrometrics/psychrolib/pull/40. Please check and let me know so that I can merge it in. I would wait to release a new version when PR https://github.com/psychrometrics/psychrolib/pull/39 is also merged in.

@DJGosnell please can you propagate the same fix in https://github.com/psychrometrics/psychrolib/pull/40 to your contribution.

didierthevenard commented:
Hi David:

No issues found, thanks. Can I let you propagate the .bas changes to the 
test_psychrolib_ip.xlsm and test_psychrolib_si.xlsm files too?

Thanks, Didier.

On 2019-10-19 5:29 a.m., dmey wrote:
>
> @didierthevenard <https://github.com/didierthevenard> well spotted! 
> You are correct -- the coefficient from the first derivative should 
> have been 3 and not 2. I have fixed this in #40 
> <https://github.com/psychrometrics/psychrolib/pull/40>. Please check 
> and let me know so that I can merge it in. I would wait to release a 
> new version when PR #39 
> <https://github.com/psychrometrics/psychrolib/pull/39> is also merged in.
>
> @DJGosnell <https://github.com/DJGosnell> please can you propagate the 
> same fix in #40 <https://github.com/psychrometrics/psychrolib/pull/40> 
> to your contribution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/psychrometrics/psychrolib/pull/39?email_source=notifications&email_token=ACXKO7JMJK2MS66TUKPIJ7TQPLHQDA5CNFSM4I6ILT7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBXJYTY#issuecomment-544119887>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ACXKO7NKWX6YUBEUM2Z4W63QPLHQDANCNFSM4I6ILT7A>.
>

dmey commented:
Hi Didier,

Thanks and sure!

Cheers,
David

From: didierthevenard <notifications@github.com>
Sent: 19 October 2019 21:43
To: psychrometrics/psychrolib <psychrolib@noreply.github.com>
Cc: David Meyer <D.Meyer@pgr.reading.ac.uk>; Mention <mention@noreply.github.com>
Subject: Re: [psychrometrics/psychrolib] Added support for C#, VB & F# Languages with .NET Standard 1.0 (#39)

Hi David:

No issues found, thanks. Can I let you propagate the .bas changes to the
test_psychrolib_ip.xlsm and test_psychrolib_si.xlsm files too?

Thanks, Didier.

On 2019-10-19 5:29 a.m., dmey wrote:
>
> @didierthevenard <https://github.com/didierthevenard> well spotted!
> You are correct -- the coefficient from the first derivative should
> have been 3 and not 2. I have fixed this in #40
> <https://github.com/psychrometrics/psychrolib/pull/40>. Please check
> and let me know so that I can merge it in. I would wait to release a
> new version when PR #39
> <https://github.com/psychrometrics/psychrolib/pull/39> is also merged in.
>
> @DJGosnell <https://github.com/DJGosnell> please can you propagate the
> same fix in #40 <https://github.com/psychrometrics/psychrolib/pull/40>
> to your contribution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psychrometrics/psychrolib/pull/39?email_source=notifications&email_token=ACXKO7JMJK2MS66TUKPIJ7TQPLHQDA5CNFSM4I6ILT7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBXJYTY#issuecomment-544119887>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACXKO7NKWX6YUBEUM2Z4W63QPLHQDANCNFSM4I6ILT7A>.
>


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/psychrometrics/psychrolib/pull/39?email_source=notifications&email_token=AGJAYW7J7BAUPF2H2AO7ZILQPNWO3A5CNFSM4I6ILT7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBX4KAA#issuecomment-544195840>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGJAYWY2YK3YOANXYOMEF33QPNWO3ANCNFSM4I6ILT7A>.

DJGosnell commented:
@didierthevenard
1. Fixed.
2. The first UnitSystem is the enum type defined lower in the psychrolib.cs file and the second is the property name.  It can be a bit confusing while looking at the code, but it really makes sense in usage.  It is also considered standard C# practice when the types and properties are so closely coupled.  [See "public Color Color" example.](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members)
``` csharp
var psy = new Psychrometrics(UnitSystem.IP);
// Change to SI unit system.
psy.UnitSystem = UnitSystem.SI;
```
3. It is not possible to initialize the class without specifying the unit system in the constructor since there is not a constructor without a parameter.
4. Yes, the reason is that when I was porting, I forgot to go back and fix the variables afterword to standard C# naming conventions.  These referenced items have been fixed and the rest of the variables have been unified in their naming.
5. Yes, the ampersand is not allowed.  It has to be converted to the HTML notation with is &amp;.  While the code compiles if not changed to ```&amp;```, the automatic parsing will fail on the comment and will display nothing in common code editors (Visual Studio).  So, I can remove the ```&amp;``` and make it readable in the source, or leave it and allow for readability in end usage.  I would suggest leaving with ```&amp;``` as the package manager (NuGet) will also fail to parse the comments.
6. Once again, this was a porting oversight.  All headers have been re-included and unified the formatting.
7. I have updated the code in regards to this.

didierthevenard commented:
@DJGosnell 
Thanks for the detailed explanations, all clear now. I am okay with leaving & with amp; since it is what works best.
